File: BIMESTRE_2/PROGRAMACION_IMPERATIVA/MODULO_02/C06_CONTROL_DE_FLUJO/PRACTICA/ex05_Paracaidas.js
const prompt = require("prompt-sync")({ sigint: true });

// ============================================================================
// 🪂 EJERCICIO 5: Abrir Paracaídas
// ============================================================================

function abrirParacaidas(velocidad, altura) {
    const velocidadValida = velocidad < 1000;
    const alturaValida = altura >= 2000 && altura < 3000;
    
    if (velocidadValida && alturaValida) {
        return "✅ Abrir paracaídas AHORA";
    } else {
        return "⚠️ NO abrir paracaídas - Condiciones no seguras";
    }
}

console.log("\n");
console.log("████████████████████████████████████████████████████████████████████");
console.log("█                                                                  █");
console.log("█                       👨‍💻 USER TESTING 👩‍💻                     █");
console.log("█                                                                  █");
console.log("████████████████████████████████████████████████████████████████████");
console.log("\n");

let velocidad = prompt(">>> Ingrese Velocidad : ");
let altura = prompt(">>> Ingrese Altura : ");

console.log(abrirParacaidas(velocidad, altura));

// ============================================================================
// 🧪 FRAMEWORK DE TESTING PROFESIONAL
// ============================================================================

class TestSuite {
    constructor(name) {
        this.name = name; 
        this.passed = 0;
        this.failed = 0;
        this.total = 0;
        this.result = [];
    }

    assert(testName, actual, expected, comparator = (a, e) => a === e) {
        this.total++;
        const passed = comparator(actual,expected);

        if(passed) {
            this.passed++;
        } else {
            this.failed++;
        }

        this.result.push({
            name : testName,
            passed,
            actual,
            expected
        });
    }

    assertNumeric(testName, actual, expected, tolerance = 0.1) {
        this.assert(testName, actual, expected, (a, e) => Math.abs(a-e) < tolerance);
    }

    report() {
        console.log("\n" + "=".repeat(70));
        console.log(` 🧪 TEST SUITE : ${this.name}`);
        console.log("=".repeat(70));

        this.result.forEach(result => {
            const icon = result.passed ? "\n✅" : "\n❌";
            const status = result.passed ? "PASS" : "FAIL";
            console.log(`${icon} ${status} | ${result.name}`);
            if (!result.passed) {
                console.log(`   Expected: ${JSON.stringify(result.expected)}`);
                console.log(`   Got:      ${JSON.stringify(result.actual)}`);
            }else{
                console.log(`   Expected: ${JSON.stringify(result.expected)}`);
                console.log(`   Got     : ${JSON.stringify(result.actual)}`);
            }
        });
        
        console.log("=".repeat(70));
        const percentage = ((this.passed / this.total) * 100).toFixed(1);
        const statusIcon = this.failed === 0 ? "🏆" : "⚠️";
        console.log(`${statusIcon} RESULTADO: ${this.passed}/${this.total} PASSED (${percentage}%)`);
        console.log("=".repeat(70) + "\n");
    }
}

// ============================================================================
// 🎯 BATERÍAS DE PRUEBAS
// ============================================================================

function testParacaidas() {
    const suite = new TestSuite("Abrir Paracaídas");
    
    // Condiciones perfectas
    suite.assert("Velocidad 800, Altura 2500 - ABRIR", 
        abrirParacaidas(800, 2500), "✅ Abrir paracaídas AHORA");
    suite.assert("Velocidad 500, Altura 2000 - ABRIR", 
        abrirParacaidas(500, 2000), "✅ Abrir paracaídas AHORA");
    suite.assert("Velocidad 999, Altura 2999 - ABRIR", 
        abrirParacaidas(999, 2999), "✅ Abrir paracaídas AHORA");
    suite.assert("Velocidad 100, Altura 2200 - ABRIR", 
        abrirParacaidas(100, 2200), "✅ Abrir paracaídas AHORA");
    
    // Velocidad muy alta
    suite.assert("Velocidad 1000, Altura 2500 - NO ABRIR", 
        abrirParacaidas(1000, 2500), "⚠️ NO abrir paracaídas - Condiciones no seguras");
    suite.assert("Velocidad 1500, Altura 2500 - NO ABRIR", 
        abrirParacaidas(1500, 2500), "⚠️ NO abrir paracaídas - Condiciones no seguras");
    suite.assert("Velocidad 2000, Altura 2500 - NO ABRIR", 
        abrirParacaidas(2000, 2500), "⚠️ NO abrir paracaídas - Condiciones no seguras");
    
    // Altura muy baja
    suite.assert("Velocidad 500, Altura 1999 - NO ABRIR", 
        abrirParacaidas(500, 1999), "⚠️ NO abrir paracaídas - Condiciones no seguras");
    suite.assert("Velocidad 500, Altura 1000 - NO ABRIR", 
        abrirParacaidas(500, 1000), "⚠️ NO abrir paracaídas - Condiciones no seguras");
    suite.assert("Velocidad 500, Altura 500 - NO ABRIR", 
        abrirParacaidas(500, 500), "⚠️ NO abrir paracaídas - Condiciones no seguras");
    
    // Altura muy alta
    suite.assert("Velocidad 500, Altura 3000 - NO ABRIR", 
        abrirParacaidas(500, 3000), "⚠️ NO abrir paracaídas - Condiciones no seguras");
    suite.assert("Velocidad 500, Altura 3500 - NO ABRIR", 
        abrirParacaidas(500, 3500), "⚠️ NO abrir paracaídas - Condiciones no seguras");
    suite.assert("Velocidad 500, Altura 5000 - NO ABRIR", 
        abrirParacaidas(500, 5000), "⚠️ NO abrir paracaídas - Condiciones no seguras");
    
    // Ambas condiciones malas
    suite.assert("Velocidad 1200, Altura 1500 - NO ABRIR", 
        abrirParacaidas(1200, 1500), "⚠️ NO abrir paracaídas - Condiciones no seguras");
    suite.assert("Velocidad 1500, Altura 3500 - NO ABRIR", 
        abrirParacaidas(1500, 3500), "⚠️ NO abrir paracaídas - Condiciones no seguras");
    
    // Casos límite exactos
    suite.assert("Velocidad 999, Altura 2000 - ABRIR (límites)", 
        abrirParacaidas(999, 2000), "✅ Abrir paracaídas AHORA");
    suite.assert("Velocidad 999, Altura 2999 - ABRIR (límites)", 
        abrirParacaidas(999, 2999), "✅ Abrir paracaídas AHORA");
    suite.assert("Velocidad 0, Altura 2500 - ABRIR", 
        abrirParacaidas(0, 2500), "✅ Abrir paracaídas AHORA");
    suite.assert("Velocidad 700, Altura 2750 - ABRIR", 
        abrirParacaidas(700, 2750), "✅ Abrir paracaídas AHORA");
    suite.assert("Velocidad 300, Altura 2100 - ABRIR", 
        abrirParacaidas(300, 2100), "✅ Abrir paracaídas AHORA");
    
    suite.report();
}

console.log("\n");
console.log("████████████████████████████████████████████████████████████████████");
console.log("█                                                                  █");
console.log("█                       🚀 UNIT TESTING 🚀                         █");
console.log("█                                                                  █");
console.log("████████████████████████████████████████████████████████████████████");
console.log("\n");

testParacaidas();

console.log("\n");
console.log("████████████████████████████████████████████████████████████████████");
console.log("█                                                                  █");
console.log("█              ✨ TODAS LAS PRUEBAS COMPLETADAS ✨                 █");
console.log("█                                                                  █");
console.log("████████████████████████████████████████████████████████████████████");
console.log("\n");

File: BIMESTRE_2/PROGRAMACION_IMPERATIVA/MODULO_02/C06_CONTROL_DE_FLUJO/PRACTICA/ex05_Paracaidas.py
# ============================================================================
# 🪂 EJERCICIO 5: Abrir Paracaídas
# ============================================================================
from test_framework import TestSuite, TestRunner

def abrir_paracaidas(velocidad: float, altura: float) -> str:
    velocidad_valida = velocidad < 1000
    altura_valida = 2000 <= altura < 3000
    
    if velocidad_valida and altura_valida:
        return "✅ Abrir paracaídas AHORA"
    else:
        return "⚠️ NO abrir paracaídas - Condiciones no seguras"
    


# ============================================================================
# 🎯 BATERÍAS DE PRUEBAS
# ============================================================================

def test_paracaidas():
    suite = TestSuite("Abrir Paracaídas")
    
    # Condiciones perfectas
    suite.assert_equal("Velocidad 800, Altura 2500 - ABRIR", 
        abrir_paracaidas(800, 2500), "✅ Abrir paracaídas AHORA")
    suite.assert_equal("Velocidad 500, Altura 2000 - ABRIR", 
        abrir_paracaidas(500, 2000), "✅ Abrir paracaídas AHORA")
    suite.assert_equal("Velocidad 999, Altura 2999 - ABRIR", 
        abrir_paracaidas(999, 2999), "✅ Abrir paracaídas AHORA")
    suite.assert_equal("Velocidad 100, Altura 2200 - ABRIR", 
        abrir_paracaidas(100, 2200), "✅ Abrir paracaídas AHORA")
    
    # Velocidad muy alta
    suite.assert_equal("Velocidad 1000, Altura 2500 - NO ABRIR", 
        abrir_paracaidas(1000, 2500), "⚠️ NO abrir paracaídas - Condiciones no seguras")
    suite.assert_equal("Velocidad 1500, Altura 2500 - NO ABRIR", 
        abrir_paracaidas(1500, 2500), "⚠️ NO abrir paracaídas - Condiciones no seguras")
    suite.assert_equal("Velocidad 2000, Altura 2500 - NO ABRIR", 
        abrir_paracaidas(2000, 2500), "⚠️ NO abrir paracaídas - Condiciones no seguras")
    
    # Altura muy baja
    suite.assert_equal("Velocidad 500, Altura 1999 - NO ABRIR", 
        abrir_paracaidas(500, 1999), "⚠️ NO abrir paracaídas - Condiciones no seguras")
    suite.assert_equal("Velocidad 500, Altura 1000 - NO ABRIR", 
        abrir_paracaidas(500, 1000), "⚠️ NO abrir paracaídas - Condiciones no seguras")
    suite.assert_equal("Velocidad 500, Altura 500 - NO ABRIR", 
        abrir_paracaidas(500, 500), "⚠️ NO abrir paracaídas - Condiciones no seguras")
    
    # Altura muy alta
    suite.assert_equal("Velocidad 500, Altura 3000 - NO ABRIR", 
        abrir_paracaidas(500, 3000), "⚠️ NO abrir paracaídas - Condiciones no seguras")
    suite.assert_equal("Velocidad 500, Altura 3500 - NO ABRIR", 
        abrir_paracaidas(500, 3500), "⚠️ NO abrir paracaídas - Condiciones no seguras")
    suite.assert_equal("Velocidad 500, Altura 5000 - NO ABRIR", 
        abrir_paracaidas(500, 5000), "⚠️ NO abrir paracaídas - Condiciones no seguras")
    
    # Ambas condiciones malas
    suite.assert_equal("Velocidad 1200, Altura 1500 - NO ABRIR", 
        abrir_paracaidas(1200, 1500), "⚠️ NO abrir paracaídas - Condiciones no seguras")
    suite.assert_equal("Velocidad 1500, Altura 3500 - NO ABRIR", 
        abrir_paracaidas(1500, 3500), "⚠️ NO abrir paracaídas - Condiciones no seguras")
    
    # Casos límite exactos
    suite.assert_equal("Velocidad 999, Altura 2000 - ABRIR (límites)", 
        abrir_paracaidas(999, 2000), "✅ Abrir paracaídas AHORA")
    suite.assert_equal("Velocidad 999, Altura 2999 - ABRIR (límites)", 
        abrir_paracaidas(999, 2999), "✅ Abrir paracaídas AHORA")
    suite.assert_equal("Velocidad 0, Altura 2500 - ABRIR", 
        abrir_paracaidas(0, 2500), "✅ Abrir paracaídas AHORA")
    suite.assert_equal("Velocidad 700, Altura 2750 - ABRIR", 
        abrir_paracaidas(700, 2750), "✅ Abrir paracaídas AHORA")
    suite.assert_equal("Velocidad 300, Altura 2100 - ABRIR", 
        abrir_paracaidas(300, 2100), "✅ Abrir paracaídas AHORA")
    
    suite.report()

def run_all_tests():
    print("\n" + "█" * 70)
    print("█" + " " * 68 + "█")
    print("█" + "  🚀 SUITE COMPLETA DE TESTING - PYTHON 🚀  ".center(66) + "█")
    print("█" + " " * 68 + "█")
    print("█" * 70 + "\n")

    test_paracaidas()

def test_user():
    print("\n" + "█" * 70)
    print("█" + " " * 68 + "█")
    print("█" + "  🚀 USER TESTING - PYTHON 🚀  ".center(66) + "█")
    print("█" + " " * 68 + "█")
    print("█" * 70 + "\n")

    velocidad = int(input("Ingrese Velocidad : "))
    altura = int(input("Ingrese Altura : "))
    print(abrir_paracaidas(velocidad,altura))

if __name__=="__main__":
    test_user()
    run_all_tests()